data_steward/cdr_cleaner/cleaning_rules/base_cleaning_rule.py

```diff
@@ -53,6 +53,26 @@ def setup_rule(self, *args, **keyword_args):
         """
         pass
 
+    @abstractmethod
+    def setup_validation(self, *args, **keyword_args):
+        """
+        Run required steps for validation setup
+
+        Method to run .  For example, if your class requires loading a static
+        table, that load operation should be defined here.  It SHOULD NOT BE
+        defined as part of get_query_specs().
+        """
+        pass
+
+    @abstractmethod
+    def get_affected_tables(self, *args, **keyword_args):
+        """
+        Method to get tables that will be modified by the cleaning rule.
+
+        :return: a list of table names that will be affected by this cleaning rule
+        """
+        pass
```

Comment:
I had this as a method because some times we need to implement some logic to see which tables will be affected by it. Most of the cleaning rules are dependant on data we get from sites and rdr if the data changes the tables that gets affected might change. SO I thought getting the tables dynamically will give all the tables for that state of data.

Reply:
So, you think there are some cases where the list of returned tables may change?  Is this because data quality has improved and the rows aren't cleaned by a cleaning query?  I'm trying to think of an instance where this would dynamically change.  From my perspective, if a table could be changed by a cleaning rule (meaning it is checked by the rule whether or not an actual change is applied), I would want to know which tables it is checking.  Can you give me an example of the dynamic perspective you are thinking about?

Reply:
consider tables like Specimen table we don't actually care about this table right now even if the data is present or not. If we manually add the tables list we might miss this table in the cleaning rule. but if we are getting the table names dynamically using fields or another logic we will not miss this table and we don't need to manually add the table to the list when the all the sites start sending data and we actually need to apply cleaning rules to this table.

Reply:
Maybe this means we need a `setter` method to validate the value before setting the attribute and the `getter` would still return the value.

Reply:
@ksdkalluri My former Java days are leaking through.  I should have implemented the getters/setters as python properties, similar to this [SO example](https://stackoverflow.com/a/47507855/8132160).  Getting back to why the setter.  Using a setter allows the value for the variable to be validated before it is set instead of when you're getting ready to return the value.  It allows you to raise errors earlier and allows the user to set the value and have that value checked.  I'd be happy to discuss this with you if you would like that.